swift/apple/FirezoneKit/Sources/FirezoneKit/Views/UpdateNotification.swift

```diff
@@ -0,0 +1,245 @@
+//
+//  UpdateNotification.swift
+//  (c) 2024 Firezone, Inc.
+//  LICENSE: Apache-2.0
+//
+
+// Note: it should be easy to expand this module to iOS
+#if os(macOS)
+import Foundation
+import UserNotifications
+import Cocoa
+
+struct SemanticVersion: Decodable, Encodable, Comparable {
+  let major: Int
+  let minor: Int
+  let patch: Int
+
+  init(major: Int, minor: Int, patch: Int) {
+      self.major = major
+      self.minor = minor
+      self.patch = patch
+  }
+
+  // This doesn't conform to the full semver spec but it's enough for our use-case
+  private static func parse(versionString: String) -> (major: Int, minor: Int, patch: Int)? {
+      let components = versionString.split(separator: ".")
+      guard components.count == 3,
+            let major = Int(components[0]),
+            let minor = Int(components[1]),
+            let patch = Int(components[2]) else {
+          return nil
+      }
+      return (major, minor, patch)
+  }
+
+  init(from decoder: Decoder) throws {
+      let container = try decoder.singleValueContainer()
+      let versionString = try container.decode(String.self)
+
+      guard let parsed = SemanticVersion.parse(versionString: versionString) else {
+          throw DecodingError.dataCorruptedError(in: container,
```

Comment:
Hmm, not sure if this is good. I'd lean towards using an existing semver library: https://swiftpackageindex.com/search?query=semver

After all, we may use some feature of semver in the future and then it wouldn't be nice if the update checker is what blows up.

Reply:
there didn't seem to be any widely used semver library in swift, which one would you suggest using?

Reply:
I haven't looked into them in detail. It is not a very big risk so anyone with a decent API will do I'd say?

Reply:
This one seems good: https://swiftpackageindex.com/sersoft-gmbh/semver

Recent activity and quite old.

Reply:
Currently we rely on 0 external dependencies in the Swift codebase. I went through the effort to do so because I discovered the version of ZIPFoundation we used had a critical vuln that wasn't caught by dependabot.

I might suggest we err more on the side of in-house or vendoring code for Swift because

- the swift package ecosystem isn't as mature or widely used as, say Rust or Elixir
- Dependabot has shoddy support for it and has issues bumping them. See https://github.com/firezone/firezone/pull/4204

Reply:
Makes sense, no dependencies is also a worthwhile goal. Let's make sure the semver parsing can handle weird input and doesn't crash anything.

Reply:
Right now it shouldn't crash but I will add some tests.

I could just add a full implementation also is just slightly more involved but chatgpt is quite good for this task.